nav2_ceres_costaware_smoother/include/nav2_ceres_costaware_smoother/smoother_cost_function.hpp

```diff
@@ -0,0 +1,252 @@
+    const double & weight,
+    const Eigen::Matrix<T, 2, 1> & pt,
+    const Eigen::Matrix<T, 2, 1> & pt_p,
+    const Eigen::Matrix<T, 2, 1> & pt_m,
+    T & r) const
+  {
+    Eigen::Matrix<T, 2, 1> d_pt_p = pt_p - pt;
+    Eigen::Matrix<T, 2, 1> d_pt_m = pt - pt_m;
+    Eigen::Matrix<T, 2, 1> d_pt_diff = _next_to_last_length_ratio * d_pt_p - d_pt_m;
+    r += (T)weight * d_pt_diff.dot(d_pt_diff);    // objective function value
+  }
+
+  /**
+   * @brief Cost function term for maximum curved paths
+   * @param weight Weight to apply to function
+   * @param pt Point Xi for evaluation
+   * @param pt_p Point Xi+1 for calculating Xi's cost
+   * @param pt_m Point Xi-1 for calculating Xi's cost
+   * @param curvature_params A struct to cache computations for the jacobian to use
+   * @param r Residual (cost) of term
+   */
+  template<typename T>
+  inline void addCurvatureResidual(
+    const double & weight,
+    const Eigen::Matrix<T, 2, 1> & pt,
+    const Eigen::Matrix<T, 2, 1> & pt_p,
+    const Eigen::Matrix<T, 2, 1> & pt_m,
+    T & r) const
+  {
+    Eigen::Matrix<T, 2, 1> center = arcCenter(
+      pt_m, pt, pt_p,
+      _next_to_last_length_ratio < 0 ? -1 : 1);
+    if (ceres::IsInfinite(center[0])) {
+      return;
+    }
+    T turning_rad = (pt - center).norm();
+    T ki_minus_kmax = (T)1.0 / turning_rad - _params.max_curvature;
+
+    if (ki_minus_kmax <= (T)EPSILON) {
+      return;
+    }
+
+    r += (T)weight * ki_minus_kmax * ki_minus_kmax;  // objective function value
+  }
+
+  /**
+   * @brief Cost function derivative term for steering away changes in pose
+   * @param weight Weight to apply to function
+   * @param xi Point Xi for evaluation
+   * @param xi_original original point Xi for evaluation
+   * @param r Residual (cost) of term
+   */
+  template<typename T>
+  inline void addDistanceResidual(
+    const double & weight,
+    const Eigen::Matrix<T, 2, 1> & xi,
+    const Eigen::Matrix<T, 2, 1> & xi_original,
+    T & r) const
+  {
+    r += (T)weight * (xi - xi_original).squaredNorm();  // objective function value
+  }
+
+  /**
+   * @brief Cost function term for steering away from costs
+   * @param weight Weight to apply to function
+   * @param value Point Xi's cost'
+   * @param params computed values to reduce overhead
+   * @param r Residual (cost) of term
+   */
+  template<typename T>
+  inline void addCostResidual(
+    const double & weight,
+    const Eigen::Matrix<T, 2, 1> & pt,
+    const Eigen::Matrix<T, 2, 1> & pt_p1,
+    const Eigen::Matrix<T, 2, 1> & pt_m1,
+    T & r) const
+  {
+    if (_params.cost_check_points.empty()) {
+      Eigen::Matrix<T, 2,
+        1> interp_pos = (pt - _costmap_origin.template cast<T>()) / (T)_costmap_resolution;
+      T value;
+      _costmap_interpolator->Evaluate(interp_pos[1] - (T)0.5, interp_pos[0] - (T)0.5, &value);
+      r += (T)weight * value * value;  // objective function value
+    } else {
+      Eigen::Matrix<T, 2, 1> dir = tangentDir(
```

Comment:
Can you explain a bit what's going on with these "cost check points"?

Reply:
It is a compensation for not using a `footprint` polygon. Simply said: the robot can pick the cost from other places than its `base_link`. E.g. a diff drive robot with 2 actuated wheels in front and 2 caster wheels in back has its `base_link` frame between the front wheels, however an effective cost center is in the middle of the robot.

However (as noted [here](https://github.com/robotechvision/navigation2/blob/bd4ec4c87eb0c4383a35736171c0841b4dda84f9/nav2_ceres_costaware_smoother/README.md?plain=1#L25)), it's quite hard to make profit of it in a reasonable time, but since the "reasonable time" is subjective (depends on use, robot, CPU, situation, other parameters), I think it's appropriate to keep the option. The parameter could be marked as `experimental.cost_check_points` if the note is not expressive enough.

Reply:
I suppose my question is: if given a circular robot, obviously you can just use the center point and check if its >= inscribed cost. For a non-circular robot, we could make the smoother server give a footprint subscriber object to the smoother algorithm so it can directly access the footprint itself. 

Would it not be more correct to use that, or is this setup because using the footprint itself is too expensive? If the case is that its too expensive to check each point, I can understand that and it would make sense to add a `simplified_footprint_polygon` parameter or something to specify that with the note. 

Also, why weight different points differently? Why not simply take the maximum cost of the points sampled of the footprint?

Reply:
> Rename to simplified footprint, if I'm correctly understanding that this is what this is

It is more of footprint centroid than the actual footprint boundary points.

> Use normal footprint given by the smoother server, unless that's too computational

I wanted to let users pick the point (or theoretically multiple points in case of a very bizarre robot shape), but in practice we could simplify the usage and compute centroid of the footprint polygon.

> Why not use the max cost of all the points, versus weighting them?

If we choose to compute the centroid ourselves, then this is irrelevant because there would be just one point. If not, then I would go with the weighted sum. The reason is simply a parable to physics: since the obstacle repelling can be understood as a pressure applied on a part of the robot, the pressure results in a force (higher force for larger parts - that's why the weights are there) and `sum` is used instead of `max` when applying multiple forces on an object. Note: the forces are applied on different points, this is not included in this parable, but is "detected and considered" by the optimizer (a separate residual field for each centroid would maybe help the optimizer, but frankly, I haven't paid too much effort in tuning the optimizer for these bizarre-shaped multi-centroid robots)

Reply:
Ah I see, so its more like below, where the footprint is the polygon, the points are the check points, and the radius is the inflation lethal radius: 

![image](https://user-images.githubusercontent.com/14944147/161125520-4b18d6eb-7150-4a32-b055-9c89bbf1de17.png)

Conceptually if it was less extreme, you could use some single non-`(0,0)` point that is the centroid of the polygon, but I think in practice if the center of the robot isn't good (circular, square), you probably would need to be working with a few points (rectangular, odd shaped) to deal with the edges. 

I think then the name / process makes sense. You do a good job warning the user about potential side effects but it might be nice to add a sentence or two about what these represent more explicitly. For example, I'd remove the word "footprint" from any explanation and add a blob about how these are a minimal set of points around the robot to check for collision at. A diagram would be illustrative.

If we sum them all would this term blow out all the others, relative to point costs? Though I agree with your explanation that max probably doesn't do the job well. Would averaging the weighted costs of multiple points have the desired outcome? That may or may not help with the local minima issue you describe. Just a suggestion to try, if it doesn't work as well, we can stick with this.

Reply:
Yes, it's exactly like you understand it now. I'm removing the "footprint" word from the comment and adding a link to the configuration where a detailed clarification with a diagram will reside. Btw your diagram made me realize that even a (not so bizarre) rectangular shape can yield significantly from having multiple cost check points. Test result was improved by 1.6% when using 2 points (front/back) instead of one in the middle. Thank you!

An average is just a sum divided by a constant, so it will not solve the problem (however using a faster `SPARSE_NORMAL_CHOLESKY` solver enables the cost check points to give good results in < 150ms time, so it is less of a problem now). But if you consider `avg` to be better here than `sum`, I have no problem to change it. Currently when you use e.g. `w_cost: 0.02` and three cost check points with weights 1, 2 and 1, the total cost weight is effectively `(1+2+1)*0.02 = 0.08`. Using `avg` (i.e. automatically normalizing weights so that their sum equals 1), could separate the problem of tuning `w_cost` against other costs and problem of "defining" a good robot shape.

Reply:
No worries, just trying to understand! Sometimes an outside perspective can help (I know it does for me).

I leave `avg` vs `max` up to your best judgement and results, I'm just offering the suggestion that it might give you the behavior you'd like but balance better with other terms. I think the tuning problem is a good point but I wouldn't want to do it if it comes at the expense of performance / path smoothness away from obstacles. If we do leave it as `max` a note about retuning may be necessary when changing the number of these points would be good.